tools/PI/DevHome.PI/Controls/AddToolControl.xaml.cs

```diff
@@ -1,24 +1,87 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
+using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.ComponentModel;
+using System.Globalization;
+using System.IO;
+using System.Linq;
 using System.Runtime.InteropServices;
+using System.Runtime.InteropServices.WindowsRuntime;
+using System.Text.RegularExpressions;
+using System.Threading.Tasks;
+using CommunityToolkit.WinUI.UI.Controls;
 using DevHome.Common.Extensions;
 using DevHome.PI.Helpers;
+using DevHome.PI.Models;
+using IWshRuntimeLibrary;
 using Microsoft.UI.Xaml;
 using Microsoft.UI.Xaml.Controls;
+using Microsoft.UI.Xaml.Media.Imaging;
+using Serilog;
+using Windows.ApplicationModel;
+using Windows.ApplicationModel.Core;
+using Windows.Foundation;
+using Windows.Graphics.Imaging;
+using Windows.Management.Deployment;
+using Windows.Storage;
+using Windows.Storage.Streams;
 using Windows.Win32;
 using Windows.Win32.UI.Controls.Dialogs;
 
 namespace DevHome.PI.Controls;
 
-public sealed partial class AddToolControl : UserControl
+public sealed partial class AddToolControl : UserControl, INotifyPropertyChanged
 {
+    private static readonly ILogger _log = Log.ForContext("SourceContext", nameof(AddToolControl));
+
     private readonly string invalidToolInfo = CommonHelper.GetLocalizedString("InvalidToolInfoMessage");
     private readonly string messageCloseText = CommonHelper.GetLocalizedString("MessageCloseText");
 
+    // We have 3 sets of operations, and we arbitrarily divide the progress timing into 3 equal segments.
+    private const int ShortcutProcessingEndIndex = 33;
+    private const int PackageProcessingEndIndex = 67;
+
+    private InstalledAppInfo? selectedApp;
+    private List<string>? shortcuts;
+    private List<Package>? packages;
+    private List<InstalledAppInfo> allApps = [];
+    private int itemCount;
+
+    public ObservableCollection<InstalledAppInfo> SortedApps { get; set; } = [];
+
+    public event PropertyChangedEventHandler? PropertyChanged;
+
+    private bool isLoading;
+
+    public bool IsLoading
+    {
+        get => isLoading;
+        set
+        {
+            isLoading = value;
+            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(IsLoading)));
+        }
+    }
+
+    private int progressPercentage;
+
+    public int ProgressPercentage
+    {
+        get => progressPercentage;
+        set
+        {
+            progressPercentage = value;
+            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(ProgressPercentage)));
+        }
+    }
```

Comment:
Any reason this can't be
```suggestion
    [ObservableProperty]
    private bool _isLoading;
    
    [ObservableProperty]
    private int _progressPercentage;
```

Reply:
Following this up.  All Observable properties should be moved to the ViewModel whenever possible.

Reply:
This is a custom UserControl. In order to use ObservableProperty, I'd have to derive from ObservableObject. Can't have multiple base classes.

Reply:
I wonder if, in this case since a view model isn't being used, it's better to have the code-behind just modify the controls directly vs having the controls bind to a data source. Turns into a lot less code... I don't feel strongly about this though.